### PR TITLE
Refactoring using `Forwardable` and `DelegateClass`.

### PR DESCRIPTION
**Pull request by reitermarkus**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Simplify some methods and classes, as well as add an `attr_predicate` helper.

### CODE REVIEW

File: Library/Homebrew/brew.sh [hunk 1]
```diff
@@ -345,5 +345,5 @@ else
 
   # Unshift command back into argument list (unless argument list was empty).
   [[ "$HOMEBREW_ARG_COUNT" -gt 0 ]] && set -- "$HOMEBREW_COMMAND" "$@"
-  { update-preinstall; exec "$HOMEBREW_RUBY_PATH" -W0 "$HOMEBREW_LIBRARY/Homebrew/brew.rb" "$@"; }
+  { update-preinstall; exec "$HOMEBREW_RUBY_PATH" "$HOMEBREW_LIBRARY/Homebrew/brew.rb" "$@"; }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/brew.sh]: What's this change for?

File: Library/Homebrew/extend/refinements.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
+# Disable refinement warning.
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/refinements.rb]: Why is there a warning?

File: Library/Homebrew/extend/refinements.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  old_verbose = $VERBOSE
+  $VERBOSE = nil
+  using Module.new
+ensure
+  $VERBOSE = old_verbose
+end
+
+module Refinements
+  refine Module do
+    def attr_predicate(*attrs)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/refinements.rb]: What's the advantage of this over e.g. including a module?

Comment (alyssais) [inline Library/Homebrew/extend/refinements.rb]: I'm also not sure a refinement is necessary in this case.

Comment (reitermarkus) [inline Library/Homebrew/extend/refinements.rb]: Well, first I added this to the `extend/module` monkey-patch. Then I thought, let's not add another monkey-patch, let's try refinements! Something as primitive as using a simple Module didn't even cross my mind. 😄

File: Library/Homebrew/extend/predicable.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+module Predicable
```
Comment (alyssais) [inline Library/Homebrew/extend/predicable.rb]: Could this name be more descriptive?

File: Library/Homebrew/extend/predicable.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+  def attr_predicate(*attrs)
+    attrs.each do |attr|
+      define_method attr do
+        instance_variable_get("@#{attr.to_s.sub(/\?$/, "")}") == true
```
Comment (alyssais) [inline Library/Homebrew/extend/predicable.rb]: I think this should take names without the `?` so it's consistent with `attr_writer`.

Comment (reitermarkus) [inline Library/Homebrew/extend/predicable.rb]: I made it this way so it‘s easier to grep for predicates.

File: Library/Homebrew/extend/predicable.rb (hunk 4 above)
Comment (reitermarkus) [inline Library/Homebrew/extend/predicable.rb]: Suggestions? I think it fits in nicely with other `*able`s.

Comment (MikeMcQuaid) [inline Library/Homebrew/extend/predicable.rb]: `AttrPredicate` 😉

File: Library/Homebrew/extend/predicable.rb (hunk 5 above)
Comment (alyssais) [inline Library/Homebrew/extend/predicable.rb]: Consistency with Ruby is more important, IMO, but it looks like I'm in the minority.

**Pull request merged**